src/main/scala/com/github/andyglow/xml/diff/XmlDiffComputer.scala

```diff
@@ -62,67 +62,76 @@ private[diff] object XmlDiffComputer {
     }
   }
 
-  private def compute(left: Seq[xml.Node], right: Seq[xml.Node])(v: (Option[xml.Node], xml.Node) => XmlDiff): (XmlDiff, Seq[xml.Node]) = {
-    def contains(one: xml.Node, all: Seq[xml.Node]): (XmlDiff, Seq[xml.Node]) = {
+  private def compute(left: Seq[xml.Node], right: Seq[xml.Node])(v: (Option[xml.Node], xml.Node) => Eval[XmlDiff]): Eval[(XmlDiff, Seq[xml.Node])] = {
+    def contains(one: xml.Node, all: Seq[xml.Node]): Eval[(XmlDiff, Seq[xml.Node])] = {
       val lookup = Lookup(one)
       val (head, tail) = all findAndDrop lookup.apply
 
       val childrenRes = head map { x =>
-        matchChildren(one.child, x.child) flatMap { details =>
+        matchChildren(one.child, x.child).map(_.flatMap { details =>
           List(UnequalElem(one.label, details))
-        }
-      } getOrElse Eq
+        })
+      } getOrElse Eval.now(Eq)
 
-      (v(head, one) ++ childrenRes, tail)
+      for {
+        res1 <- v(head, one)
+        res2 <- childrenRes
+      } yield (res1 ++ res2, tail)
     }
 
     if (left.size == 1) {
       contains(left.head, right)
     } else if (left.size > 1) {
-      val (diff1, rest1) = contains(left.head, right)
-      val (diff2, rest2) = compute(left.tail, rest1)(v)
-      (diff1 ++ diff2, rest2)
+      for {
+        res1 <- contains(left.head, right)
+        (diff1, rest1) = res1
+        res2 <- compute(left.tail, rest1)(v)
+        (diff2, rest2) = res2
+      } yield (diff1 ++ diff2, rest2)
     } else {
-      (Eq, Seq.empty)
+      Eval.now((Eq, Seq.empty))
     }
   }
 
-  def matchChildren(e: Seq[xml.Node], a: Seq[xml.Node]): XmlDiff = {
-    val (diff1, rest) = compute(e, a) {
+  def matchChildren(e: Seq[xml.Node], a: Seq[xml.Node]): Eval[XmlDiff] = {
+    compute(e, a) {
       case (Some(that: xml.Elem), one: xml.Elem) =>
-        matchAttributes(one, that) flatMap { details =>
+        matchAttributes(one, that).map(_.flatMap { details =>
           List(UnequalElem(one.label, details))
-        }
+        })
 
       case (Some(that), one) =>
-        if (one.text.trim == that.text.trim) Eq else Neq(AbsentNode(one), RedundantNode(that))
+        Eval.now(if (one.text.trim == that.text.trim) Eq else Neq(AbsentNode(one), RedundantNode(that)))
 
-      case (None, one) => Neq(AbsentNode(one))
-    }
+      case (None, one) => Eval.now(Neq(AbsentNode(one)))
+    }.map { case (diff1, rest) =>
+      val diff2 = rest.foldLeft[XmlDiff](Eq) {
+        case (diff, that) => diff ++ Neq(RedundantNode(that))
+      }
 
-    val diff2 = rest.foldLeft[XmlDiff](Eq) {
-      case (diff, that) => diff ++ Neq(RedundantNode(that))
+      diff1 ++ diff2
     }
-
-    diff1 ++ diff2
   }
 
-  def computeMatching(e: xml.NodeSeq, a: xml.NodeSeq): XmlDiff = {
+  def computeMatching(e: xml.NodeSeq, a: xml.NodeSeq): Eval[XmlDiff] = {
     (e, a) match {
       case (e: xml.Elem, a: xml.Elem) =>
-        matchNames(e, a) ++
-        matchAttributes(e, a) ++
-        matchChildren(e.child, a.child) flatMap {
```

Comment:
I wasn't 100% sure whether `flatMap` here was being called only on the result of `matchChildren(...)` or on the whole result of `(matchNames(...) ++ matchAttributes(...) ++ matchChildren(...)`, i.e. whether `++` or `flatMap` has higher precedence.

I settled on calling `flatMap` on the whole result based on this example that proved `++` has higher precedence:

```scala
List(1) ++ List(2) ++ List(3) flatMap { i => List(i, i) }
// List(1, 1, 2, 2, 3, 3)
```

Let me know if you see any issues here though